# Proceso de ETL

## Instalación de librerias


In [277]:
%pip install -r requirements.txt

Note: you may need to restart the kernel to use updated packages.


ERROR: Could not open requirements file: [Errno 2] No such file or directory: 'requirements.txt'

[notice] A new release of pip is available: 23.1.2 -> 24.0
[notice] To update, run: python.exe -m pip install --upgrade pip


In [278]:
# !pip install pandas
# !pip install pyarrow
# !pip install gzip
# !pip install matplotlib
# !pip install folium
# !pip install numpy

## Importación de librerías

In [279]:
import pandas as pd
import pyarrow
import gzip
import warnings
warnings.filterwarnings("ignore")
import matplotlib.pyplot as plt
import folium
import numpy as np


In [280]:
path = 'path.txt'

## Funciones


In [281]:
def reemplazar_24_horas(horario:str) -> str:
    if '24' in horario:
        return '24 hours daily'
    else:
        return horario

## Extraccion

In [282]:
# Con esta funcion se lee cada linea que se toma como una variables con un string asignado
with open(path, 'r') as archivo:
    for linea in archivo.readlines():
        exec(linea.strip(), globals())

In [283]:
df = pd.read_csv(estaciones_carga, encoding='utf-8')

Realizamos un análisis preliminar del df.

In [284]:
df.head(3)

Fuel Type Code                                  Station Name  \
0            CNG          Spire - Montgomery Operations Center   
1            CNG                           PS Energy - Atlanta   
2            CNG  Metropolitan Atlanta Rapid Transit Authority   

        Street Address                            Intersection Directions  \
0     2951 Chestnut St                                                NaN   
1     340 Whitehall St  From I-7585 N, exit 91 to Central Ave, left on...   
2  2424 Piedmont Rd NE                                                NaN   

         City State    ZIP  Plus4 Station Phone Status Code  ...  \
0  Montgomery    AL  36107    NaN           NaN           E  ...   
1     Atlanta    GA  30303    NaN  770-350-3000           E  ...   
2     Atlanta    GA  30324    NaN           NaN           E  ...   

  EV Pricing (French) LPG Nozzle Types Hydrogen Pressures Hydrogen Standards  \
0                 NaN              NaN                NaN                NaN   
1                 NaN              NaN                NaN                NaN   
2                 NaN              NaN                NaN                NaN   

  CNG Fill Type Code CNG PSI CNG Vehicle Class  LNG Vehicle Class  \
0                  B    3600                MD                NaN   
1                  Q    3600                MD                NaN   
2                  Q    3000                LD                NaN   

   EV On-Site Renewable Source  Restricted Access  
0                          NaN                NaN  
1                          NaN              False  
2                          NaN                NaN  

[3 rows x 65 columns]

In [285]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 70406 entries, 0 to 70405
Data columns (total 65 columns):
 #   Column                            Non-Null Count  Dtype  
---  ------                            --------------  -----  
 0   Fuel Type Code                    70406 non-null  object 
 1   Station Name                      70406 non-null  object 
 2   Street Address                    70405 non-null  object 
 3   Intersection Directions           4492 non-null   object 
 4   City                              70406 non-null  object 
 5   State                             70406 non-null  object 
 6   ZIP                               70405 non-null  object 
 7   Plus4                             0 non-null      float64
 8   Station Phone                     65204 non-null  object 
 9   Status Code                       70406 non-null  object 
 10  Expected Date                     1218 non-null   object 
 11  Groups With Access Code           70406 non-null  object 
 12  Acce

In [286]:
df.describe()

Plus4  EV Level1 EVSE Num  EV Level2 EVSE Num  EV DC Fast Count  \
count    0.0          286.000000        54144.000000       8307.000000   
mean     NaN            3.461538            2.315289          3.576381   
std      NaN            6.548356            3.157387          4.129470   
min      NaN            1.000000            1.000000          1.000000   
25%      NaN            1.000000            2.000000          1.000000   
50%      NaN            1.000000            2.000000          2.000000   
75%      NaN            2.000000            2.000000          4.000000   
max      NaN           51.000000          311.000000         56.000000   

           Latitude     Longitude             ID  Federal Agency ID  \
count  70406.000000  70406.000000   70406.000000         955.000000   
mean      38.929309    -96.061447  148730.326066          13.905759   
std        5.628458     19.545273   54261.672118           6.767439   
min        0.000000   -159.788556      17.000000           2.000000   
25%       34.182981   -117.860563  113723.250000           8.000000   
50%       39.105989    -92.002897  165247.500000          14.000000   
75%       42.702185    -78.881493  190653.750000          17.000000   
max       64.852466     40.432278  224382.000000          60.000000   

       CNG Dispenser Num  CNG Total Compression Capacity  CNG Storage Capacity  
count        1039.000000                      708.000000            357.000000  
mean            2.731473                      873.766949          47866.787115  
std             8.333492                     1032.371466          54327.698853  
min             0.000000                        2.000000              0.000000  
25%             1.000000                      215.750000          30000.000000  
50%             2.000000                      632.000000          36000.000000  
75%             2.000000                     1100.000000          60000.000000  
max           202.000000                    10620.000000         593136.000000

In [287]:
df.shape

(70406, 65)

In [288]:
df.columns

Index(['Fuel Type Code', 'Station Name', 'Street Address',
       'Intersection Directions', 'City', 'State', 'ZIP', 'Plus4',
       'Station Phone', 'Status Code', 'Expected Date',
       'Groups With Access Code', 'Access Days Time', 'Cards Accepted',
       'BD Blends', 'NG Fill Type Code', 'NG PSI', 'EV Level1 EVSE Num',
       'EV Level2 EVSE Num', 'EV DC Fast Count', 'EV Other Info', 'EV Network',
       'EV Network Web', 'Geocode Status', 'Latitude', 'Longitude',
       'Date Last Confirmed', 'ID', 'Updated At', 'Owner Type Code',
       'Federal Agency ID', 'Federal Agency Name', 'Open Date',
       'Hydrogen Status Link', 'NG Vehicle Class', 'LPG Primary',
       'E85 Blender Pump', 'EV Connector Types', 'Country',
       'Intersection Directions (French)', 'Access Days Time (French)',
       'BD Blends (French)', 'Groups With Access Code (French)',
       'Hydrogen Is Retail', 'Access Code', 'Access Detail Code',
       'Federal Agency Code', 'Facility Type', 'CNG Dispenser N

## Transformacion de datos



#### columnas irrelevantes


Vemos, a priori, que hay muchas columnas que no nos van a servir para el proyecto, por lo que procedemos a borrarlas.

In [289]:
columnas_a_borrar = ['Station Name', 'Street Address', 'ZIP', 'Plus4', 'Station Phone', 'EV Network Web', 'Open Date', 'Hydrogen Status Link',
                     'Intersection Directions (French)', 'Access Days Time (French)', 'BD Blends (French)', 'Groups With Access Code (French)',
                     'Hydrogen Is Retail', 'EV Pricing (French)', ]

In [290]:
df = df.drop(columnas_a_borrar, axis=1)

Hay otras columnas que no sabemos. Antes de borrarlas, analizamos su contenido.

In [291]:
columnas_a_analizar = ['Groups With Access Code', 'Intersection Directions', 'Cards Accepted', 'NG PSI', 'EV Other Info', 'EV Network', 'Geocode Status',
              'Date Last Confirmed', 'Updated At', 'Owner Type Code', 'Federal Agency ID', 'Federal Agency Name',
              'NG Vehicle Class', 'LPG Primary',  'Access Code', 'Access Detail Code', 'Federal Agency Code',
              'Facility Type', 'CNG Fill Type Code', 'EV On-Site Renewable Source', 'Restricted Access']

In [292]:
for columna in columnas_a_analizar:
    print(f'\n\nColumna: {columna}\n{df[columna].unique()}\n')



Columna: Groups With Access Code
['Private' 'Public - Card key at all times' 'Private - Government only'
 'Public - Credit card at all times' 'TEMPORARILY UNAVAILABLE (Public)'
 'TEMPORARILY UNAVAILABLE (Public - Credit card at all times)'
 'Private - Fleet customers only' 'Public'
 'Public - Credit card after hours' 'TEMPORARILY UNAVAILABLE (Private)'
 'Private - Card key at all times' 'Public - Call ahead'
 'Public - Card key after hours'
 'TEMPORARILY UNAVAILABLE (Public - Credit card after hours)'
 'TEMPORARILY UNAVAILABLE (Public - Call ahead)'
 'TEMPORARILY UNAVAILABLE (Private - Government only)'
 'PLANNED - not yet accessible (Public - Credit card at all times)'
 'PLANNED - not yet accessible (Private - Government only)'
 'Public - Limited hours' 'PLANNED - not yet accessible (Public)'
 'PLANNED - not yet accessible (Public - Call ahead)'
 'Private - Call ahead' 'Private - Credit card at all times'
 'PLANNED - not yet accessible (Private)' 'Public - Fleet customers only'
 'TE

Vemos que algunas pueden llegar a servir:

'Geocode Status'

'Access Code'

Pero podemos borrar las otras.

In [293]:
columnas_a_borrar = ['Groups With Access Code', 'Intersection Directions', 'Cards Accepted', 'NG PSI', 'EV Other Info', 'EV Network',
              'Date Last Confirmed', 'Updated At', 'Owner Type Code', 'Federal Agency ID', 'Federal Agency Name',
              'NG Vehicle Class', 'LPG Primary', 'Access Detail Code', 'Federal Agency Code',
              'Facility Type', 'CNG Fill Type Code', 'EV On-Site Renewable Source', 'Restricted Access']

In [294]:
df = df.drop(columnas_a_borrar, axis=1)

La columna 'Acces Days Time' indica los horarios de atención. Vamos a analizarla:

In [295]:
df['Access Days Time'].unique()

array([nan, '24 hours daily',
       '24 hours daily; call 866-809-4869 for Clean Energy card', ...,
       '7:30am-7pm M-F; 8am-6pm Sat; Closed Sunday',
       '24 hours daily; Phone App',
       'Mon 7:00am - 6:00pm; Tue 7:00am - 6:00pm; Wed 7:00am - 6:00pm; Thu 7:00am - 6:00pm; Fri 7:00am - 6:00pm'],
      dtype=object)

In [296]:
df['Access Days Time'].nunique()

2552

Vemos que tiene muchos valores únicos. Vamos a reemplazar todos los que contengan '24' y luego volvemos a analizar.

In [297]:
df['Access Days Time'] = df['Access Days Time'].astype(str)

In [298]:
df['Access Days Time'] = df['Access Days Time'].apply(reemplazar_24_horas)

In [299]:
df['Access Days Time'].nunique()

1914

Vemos que sigue teniendo muchos valores únicos y, para el contexto de nuestro proyecto, no va a ser una columna relevante.

In [300]:
df = df.drop(['Access Days Time'], axis=1)

Verificamos valores únicos de estados y ciudades:

In [301]:
df['City'].unique()

array(['Montgomery', 'Atlanta', 'Fort Smith', ..., 'New Minas',
       'Herriman', 'Bluffdale'], dtype=object)

In [302]:
df['City'].nunique()

8278

In [303]:
df['State'].unique()

array(['AL', 'GA', 'AR', 'MA', 'NY', 'IN', 'WY', 'CT', 'LA', 'NM', 'KS',
       'ID', 'MO', 'TX', 'UT', 'MN', 'NE', 'CO', 'OR', 'OK', 'PA', 'CA',
       'FL', 'SC', 'NC', 'AZ', 'NV', 'VA', 'MD', 'WA', 'TN', 'SD', 'IA',
       'IL', 'MI', 'ND', 'NJ', 'MS', 'OH', 'MT', 'RI', 'NH', 'WV', 'WI',
       'HI', 'VT', 'AK', 'ME', 'KY', 'DC', 'DE', 'AB', 'QC', 'MB', 'SK',
       'ON', 'BC', 'NL', 'PE', 'NB', 'NS', 'NT', 'YT', 'PR', 'MX'],
      dtype=object)

Como el alcance de nuestro proyecto es la Ciudad de New York y viajes de corta distancia, no necesitamos datos de estaciones de carga de todo el país. En principio, dejamos solo los datos del estado de New York y estados limítrofes (New York, Connecticut, Massachusetts, New Jersey, Rhode Island, Pensilvania, Vermont). La idea es ir filtrando más a medida que avanza el proyecto.

In [304]:
estados_cercanos = ['NY', 'CT', 'MA', 'NJ', 'RI', 'PA', 'VT']

In [305]:
df.shape

(70406, 31)

Verificamos cuántos registros tenemos en esos estados:

In [306]:
df[df['State'].isin(estados_cercanos)].shape

(9479, 31)

De los 70406 registros solo 9479 son del estado de NY y estados limítrofes. Borramos los otros.

In [307]:
df = df.drop(df[~df['State'].isin(estados_cercanos)].index)

In [308]:
df.shape

(9479, 31)

Exploramos la columna 'Expected Date'.

In [309]:
df['Expected Date'].unique()

array([nan, '2022-07-15', '2022-05-15', '2022-06-15', '2022-03-15',
       '2022-04-15', '2022-06-13', '2022-07-31', '2022-12-01',
       '2023-02-15', '2023-05-24', '2022-07-01', '2022-07-10',
       '2023-05-15', '2022-03-30', '2021-04-30', '2022-02-01',
       '2021-07-31', '2021-04-29', '2021-06-30', '2021-11-30',
       '2022-04-14', '2022-06-30'], dtype=object)

In [310]:
df['Expected Date'].nunique()

22

Observarmos que no hay fechas esperadas posteriores a 2023, por lo que borramos la columna.

In [311]:
df = df.drop('Expected Date', axis=1)

Exploramos la columna 'Status Code'.

In [312]:
df['Status Code'].unique()

array(['E', 'T', 'P'], dtype=object)

'Status Code' es un dato temporal que se refiere al estado de la estación de carga. E: en uso; T: disponible; P: fuera de servicio. Puede servir para la carga de datos, pero solo si se actualizan con mucha frecuencia. No va a ser el caso, por lo que también la borramos.

In [313]:
df = df.drop('Status Code', axis=1)

In [314]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 9479 entries, 5 to 70395
Data columns (total 29 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   Fuel Type Code                  9479 non-null   object 
 1   City                            9479 non-null   object 
 2   State                           9479 non-null   object 
 3   BD Blends                       36 non-null     object 
 4   NG Fill Type Code               203 non-null    object 
 5   EV Level1 EVSE Num              34 non-null     float64
 6   EV Level2 EVSE Num              8197 non-null   float64
 7   EV DC Fast Count                799 non-null    float64
 8   Geocode Status                  9479 non-null   object 
 9   Latitude                        9479 non-null   float64
 10  Longitude                       9479 non-null   float64
 11  ID                              9479 non-null   int64  
 12  E85 Blender Pump                275 no

Verificamos las columnas de latitud y longitud. Sabemos por el método info que no hay valores nulos, pero chequeamos valores atípicos o errores.

El área que vamos a evaluar antes de un filtrado final está delimitada, aproximadamente, por las siguientes coordenadas:

Latitud mínima: 36

Latitud máxima: 45

Longitud mínima: -110

Longitud máxima: -69

In [315]:
df[df['Latitude'] < 36].shape

(0, 29)

In [316]:
df[df['Latitude'] > 45].shape

(3, 29)

In [317]:
df['Latitude'][df['Latitude'] > 45]

10082    45.004475
29456    45.513643
68491    45.457863
Name: Latitude, dtype: float64

Vemos que las latitudes están bien.

In [318]:
df[df['Longitude'] < -110].shape

(1, 29)

In [319]:
df['Longitude'][df['Longitude'] < -110]

29456   -122.833936
Name: Longitude, dtype: float64

In [320]:
df[['Latitude','Longitude']][df['Longitude'] < -110]

Latitude   Longitude
29456  45.513643 -122.833936

In [321]:
df[df['Longitude'] < -122]

Fuel Type Code       City State BD Blends NG Fill Type Code  \
29456           ELEC  Beaverton    RI       NaN               NaN   

       EV Level1 EVSE Num  EV Level2 EVSE Num  EV DC Fast Count  \
29456                 NaN                 2.0               NaN   

      Geocode Status   Latitude  ...  CNG Storage Capacity  \
29456            GPS  45.513643  ...                   NaN   

       LNG On-Site Renewable Source E85 Other Ethanol Blends EV Pricing  \
29456                           NaN                      NaN        NaN   

      LPG Nozzle Types Hydrogen Pressures  Hydrogen Standards CNG PSI  \
29456              NaN                NaN                 NaN     NaN   

       CNG Vehicle Class  LNG Vehicle Class  
29456                NaN                NaN  

[1 rows x 29 columns]

Vemos que el estado está bien pero no la longitud. Como no tenemos el dato exacto, borramos el registro.

In [322]:
df.shape

(9479, 29)

In [323]:
df = df.drop(df[df['Longitude'] < -122].index)

In [324]:
df.shape

(9478, 29)

In [325]:
df[df['Longitude'] > -69].shape

(2, 29)

In [326]:
df[['Latitude', 'Longitude']][df['Longitude'] > -69]

Latitude  Longitude
53136  42.868419   0.000000
69951  40.432278  40.432278

In [327]:
df[df['Longitude'] > -69]

Fuel Type Code          City State BD Blends NG Fill Type Code  \
53136           ELEC  Clifton Park    NY       NaN               NaN   
69951           ELEC    Pittsburgh    PA       NaN               NaN   

       EV Level1 EVSE Num  EV Level2 EVSE Num  EV DC Fast Count  \
53136                 NaN                 4.0               NaN   
69951                 NaN                 2.0               NaN   

      Geocode Status   Latitude  ...  CNG Storage Capacity  \
53136            GPS  42.868419  ...                   NaN   
69951            GPS  40.432278  ...                   NaN   

       LNG On-Site Renewable Source E85 Other Ethanol Blends  \
53136                           NaN                      NaN   
69951                           NaN                      NaN   

                   EV Pricing LPG Nozzle Types Hydrogen Pressures  \
53136                     NaN              NaN                NaN   
69951  Level 2: $0.49 per kWh              NaN                NaN   

       Hydrogen Standards CNG PSI  CNG Vehicle Class  LNG Vehicle Class  
53136                 NaN     NaN                NaN                NaN  
69951                 NaN     NaN                NaN                NaN  

[2 rows x 29 columns]

De nuevo, los estados están bien, pero no las coordenadas. Como no tenemos el dato exacto, borramos los registros.

In [328]:
df.shape

(9478, 29)

In [329]:
df = df.drop(df[df['Longitude'] > -69].index)

In [330]:
df.shape

(9476, 29)

Verificamos:

In [331]:
for latitud in df['Latitude']:
    if latitud < 36 or latitud > 45:
        print(latitud)

45.004475
45.45786273


In [332]:
for latitud in df['Longitude']:
    if latitud < -110 or latitud > -69:
        print(latitud)

Seguimos explorando columnas:

In [333]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 9476 entries, 5 to 70395
Data columns (total 29 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   Fuel Type Code                  9476 non-null   object 
 1   City                            9476 non-null   object 
 2   State                           9476 non-null   object 
 3   BD Blends                       36 non-null     object 
 4   NG Fill Type Code               203 non-null    object 
 5   EV Level1 EVSE Num              34 non-null     float64
 6   EV Level2 EVSE Num              8194 non-null   float64
 7   EV DC Fast Count                799 non-null    float64
 8   Geocode Status                  9476 non-null   object 
 9   Latitude                        9476 non-null   float64
 10  Longitude                       9476 non-null   float64
 11  ID                              9476 non-null   int64  
 12  E85 Blender Pump                275 no

Borramos las columnas que no son importantes para el alcance del proyecto. Dejamos algunas relacionadas con vehículos eléctricos (como precio de carga) porque pueden llegar a servir, para hacer una transformación de datos.

In [334]:
columnas_a_borrar = ['BD Blends', 'NG Fill Type Code', 'E85 Blender Pump', 'CNG On-Site Renewable Source',
                     'CNG Total Compression Capacity', 'CNG Storage Capacity', 'LNG On-Site Renewable Source',
                     'E85 Other Ethanol Blends', 'Hydrogen Pressures', 'Hydrogen Standards',
                     'CNG PSI', 'CNG Vehicle Class', 'LNG Vehicle Class']

In [335]:
df.shape

(9476, 29)

In [336]:
df = df.drop(columnas_a_borrar, axis=1)

In [337]:
df.shape

(9476, 16)

Ordenamos las columnas y borramos algunas que notamos que son irrelevantes al ir avanzando con el análisis.

In [338]:
df.columns

Index(['Fuel Type Code', 'City', 'State', 'EV Level1 EVSE Num',
       'EV Level2 EVSE Num', 'EV DC Fast Count', 'Geocode Status', 'Latitude',
       'Longitude', 'ID', 'EV Connector Types', 'Country', 'Access Code',
       'CNG Dispenser Num', 'EV Pricing', 'LPG Nozzle Types'],
      dtype='object')

In [339]:
df = df[['ID', 'Fuel Type Code', 'Access Code', 'Geocode Status', 'Latitude', 'Longitude', 'City', 'State',
         'EV Level1 EVSE Num', 'EV Level2 EVSE Num', 'EV DC Fast Count', 'EV Connector Types', 'EV Pricing',
         'CNG Dispenser Num', 'LPG Nozzle Types']]

In [340]:
df.head()

ID Fuel Type Code Access Code Geocode Status   Latitude  Longitude  \
5   81            CNG      public            GPS  42.374706 -71.026549   
6   84            CNG      public          200-8  42.393167 -71.064352   
7  108            CNG      public          200-8  40.718037 -73.932309   
8  112            CNG      public          200-8  40.645540 -73.918344   
9  124            CNG      public          200-9  40.752903 -74.005831   

          City State  EV Level1 EVSE Num  EV Level2 EVSE Num  \
5  East Boston    MA                 NaN                 NaN   
6      Everett    MA                 NaN                 NaN   
7     Brooklyn    NY                 NaN                 NaN   
8     Brooklyn    NY                 NaN                 NaN   
9     New York    NY                 NaN                 NaN   

   EV DC Fast Count EV Connector Types EV Pricing  CNG Dispenser Num  \
5               NaN                NaN        NaN                4.0   
6               NaN                NaN        NaN                1.0   
7               NaN                NaN        NaN                1.0   
8               NaN                NaN        NaN                1.0   
9               NaN                NaN        NaN                1.0   

  LPG Nozzle Types  
5              NaN  
6              NaN  
7              NaN  
8              NaN  
9              NaN

In [341]:
df.shape

(9476, 15)

In [342]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 9476 entries, 5 to 70395
Data columns (total 15 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   ID                  9476 non-null   int64  
 1   Fuel Type Code      9476 non-null   object 
 2   Access Code         9476 non-null   object 
 3   Geocode Status      9476 non-null   object 
 4   Latitude            9476 non-null   float64
 5   Longitude           9476 non-null   float64
 6   City                9476 non-null   object 
 7   State               9476 non-null   object 
 8   EV Level1 EVSE Num  34 non-null     float64
 9   EV Level2 EVSE Num  8194 non-null   float64
 10  EV DC Fast Count    799 non-null    float64
 11  EV Connector Types  8850 non-null   object 
 12  EV Pricing          2282 non-null   object 
 13  CNG Dispenser Num   131 non-null    float64
 14  LPG Nozzle Types    96 non-null     object 
dtypes: float64(6), int64(1), object(8)
memory usage: 1.2+ MB


Observamos que tenemos muchos valores nulos, pero es porque algunas columnas no aplican para todos los combustibles. Vamos a crear un df para cada tipo de combustible y analizarlo en función de esa clasificación.

In [343]:
df.to_csv(estaciones_ETL, encoding='utf-8')